### PR TITLE
Made initialize_universe parallel and fixed clippy warnings

### PR DESCRIPTION
**Pull request by ghost**
I ran cargo fmt which is why so many lines changed

### CODE REVIEW

**Comment by selfup** [comment]
Ok I will pull this down and review. Might take me a while. Have been busy lately. Thanks for the PR!

**Comment by selfup** [comment]
Second pass runs faster!

```
$ ./scripts/generate.sh 480
--------------------------------
Threads: 16
Building..
--------------------------------
Universe built!
--------------------------------
Field is Anionic
--------------------------------
Atoms: 110592000
Baryons: 26099712000
Quarks: 78299136000
--------------------------------

real    0m12.475s
user    0m0.000s
sys     0m0.015s
```

Currently all tests still pass so that's nice:

```
running 10 tests
test builder::core::neutron::it_can_match_a_single_neutron_correctly ... ok
test builder::core::neutron::it_defaults_all_neutrons_as_unknown ... ok
test builder::core::proton::it_defaults_all_protons_as_unknown ... ok
Threads: 16
Building..
Threads: 16
Building..
Threads: 16
Building..
test builder::core::proton::it_can_match_a_single_proton_correctly ... ok
test builder::core::quark::it_infers_correct_electric_charges ... ok
test builder::core::quark::it_infers_correct_strong_charges ... ok
test builder::core::quark::it_infers_the_correct_combination_of_charges ... ok
test builder::it_can_sense_the_field ... ok
test builder::it_can_begin ... ok
test builder::it_can_infer_the_charge_of_an_atom ... ok

test result: ok. 10 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

File: crates/unigen/src/builder/core/neutron.rs [hunk 1]
```diff
@@ -9,19 +9,21 @@ pub struct Neutrons {
 impl Neutrons {
     pub fn new(count: u32) -> Self {
         let mut neutrons = [NeutronData::Unknown; 118];
+        neutrons[0..(count as usize)]
+            .iter_mut()
+            .for_each(|n| *n = NeutronData::new(Neutron::new()));
 
+        /*
```
Comment (selfup) [inline crates/unigen/src/builder/core/neutron.rs]: is this slower than iter_mut? Sometimes single threaded wins. It depends 🙃

Comment (ghost) [reply crates/unigen/src/builder/core/neutron.rs]: `iter_mut` is single threaded you might be thinking of rayon's `par_iter_mut` 
but anyway from what i found from testing on rust playground using `iter_mut` seems to be able to do an inline memcpy compared to directly indexing which compiles down to a simple move compare and jump

Comment (ghost) [reply crates/unigen/src/builder/core/neutron.rs]: I just ran a test and using `iter_mut` performs maybe a few percent faster but not really noticeable.
If you don't like to `for_each` you could replace it with something like this
```
for x in &mut arr[0..end]{
    *x = value;
}
```

Comment (selfup) [reply crates/unigen/src/builder/core/neutron.rs]: Ok makes sense. I don't mind either approach. I guess I just saw `iter_mut` and assumed `par`.

Comment (selfup) [reply crates/unigen/src/builder/core/neutron.rs]: In this case you can get rid of the commented out for blocks. I like it.

**Comment by ghost** [comment]
I just did the tests and the test `builder::it_can_infer_the_charge_of_an_atom` fails however this test seems to fail even before I made this PR I get the error 
`attempt to subtract with overflow` at line 138 builder.rs

**Comment by selfup** [comment]
Interesting. I will check that out this week!

**Comment by selfup** [comment]
```
commit 5003f33a95a8efa809e742dd8ac63b0b960a54e8 (HEAD -> master, origin/master, origin/HEAD)
```

```
$ ./scripts/test.sh | grep 'failed'
    Finished release [optimized] target(s) in 0.35s
     Running target\release\deps\oxidizy-db60475d0555543a.exe
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
    Finished release [optimized] target(s) in 0.10s
     Running C:\Users\selfup\Documents\oxidizy\target\release\deps\unigen-dea8487b230ab01c.exe
test result: ok. 10 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
     Running C:\Users\selfup\Documents\oxidizy\target\release\deps\unigen-a2b46028b5bd9c5d.exe
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
   Doc-tests unigen
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

No failures here.

Let me pull down your PR and check.

**Comment by selfup** [comment]
```
$ rustc --version
rustc 1.50.0 (cb75ad5db 2021-02-10)
```

```
alias gfo="git fetch origin"
alias gk="git checkout"
```

```
gfo pull/20/head:initialize_universe_clippy
gk initialize_universe_clippy
```

```
commit 5ce08174c3f3eb4097578213e1e0534609cc9638 (HEAD -> initialize_universe_clippy)
```

```
~/Documents/oxidizy (initialize_universe_clippy)
$ ./scripts/test.sh | grep 'failed'
   Compiling unigen v0.1.0 (C:\Users\selfup\Documents\oxidizy\crates\unigen)
   Compiling oxidizy v0.1.0 (C:\Users\selfup\Documents\oxidizy)
    Finished release [optimized] target(s) in 1.76s
     Running target\release\deps\oxidizy-db60475d0555543a.exe
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
   Compiling unigen v0.1.0 (C:\Users\selfup\Documents\oxidizy\crates\unigen)
    Finished release [optimized] target(s) in 1.44s
     Running C:\Users\selfup\Documents\oxidizy\target\release\deps\unigen-dea8487b230ab01c.exe
test result: ok. 10 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
     Running C:\Users\selfup\Documents\oxidizy\target\release\deps\unigen-a2b46028b5bd9c5d.exe
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
   Doc-tests unigen
   ```

**Comment by selfup** [comment]
I checked on 1.49.x as well as 1.50.0 and no overflow issues here. I will merge!

**Comment by ghost** [comment]
the tests succeed as the script runs the tests in release mode which  doesn't do arithmetic checks.

**Comment by selfup** [comment]
Thanks for the heads up. The issue with `.sum()` is that it only works with `u32` so this could get interesting..

**Comment by ghost** [comment]
```
pub fn mutate_blocks_with_new_particles(rng: &mut rand::rngs::ThreadRng, block: &mut core::Block) {
    let (electrons, protons, neutrons, rotation): (u32, u32, u32, u8) = (
        rng.gen_range(0, 118),
        rng.gen_range(0, 118),
        rng.gen_range(0, 118),
        rng.gen_range(1, 6),
    );

    match rotation {
        1 => block.x += 1,
        2 => block.x -= 1,  // block.x = 0 as its unsigned an underflow occurs
        3 => block.y += 1,
        4 => block.y -= 1,
        5 => block.z += 1,
        6 => block.z -= 1,
        _ => (),
    }

    block.atom.electrons = electrons;
    block.atom.nucleus.baryon.protons = proton::Protons::new(protons);
    block.atom.nucleus.baryon.neutrons = neutron::Neutrons::new(neutrons);

```
should block.(x , y, z) be an i32 not u32?

**Comment by ghost** [comment]
I created a new pr which fixes this issue

**Pull request merged**